crates/scroll/evm/src/execute.rs

```diff
@@ -0,0 +1,545 @@
+        total_difficulty: U256,
+    ) -> Result<ExecuteOutput, Self::Error> {
+        let env = self.evm_env_for_block(&block.header, total_difficulty);
+        let mut evm = self.evm_config.evm_with_env(&mut self.state, env);
+
+        let mut cumulative_gas_used = 0;
+        let mut receipts = Vec::with_capacity(block.body.transactions.len());
+
+        for (sender, transaction) in block.transactions_with_sender() {
+            // The sum of the transaction’s gas limit and the gas utilized in this block prior,
+            // must be no greater than the block’s gasLimit.
+            let block_available_gas = block.header.gas_limit - cumulative_gas_used;
+            if transaction.gas_limit() > block_available_gas {
+                return Err(BlockValidationError::TransactionGasLimitMoreThanAvailableBlockGas {
+                    transaction_gas_limit: transaction.gas_limit(),
+                    block_available_gas,
+                }
+                .into())
+            }
+
+            if transaction.is_eip4844() {
+                return Err(ConsensusError::InvalidTransaction(
+                    InvalidTransactionError::Eip4844Disabled,
+                )
+                .into())
+            }
+
+            self.evm_config.fill_tx_env(evm.tx_mut(), transaction, *sender);
+            if transaction.is_l1_message() {
+                evm.context.evm.env.cfg.disable_base_fee = true; // disable base fee for l1 msg
+            } else {
+                // RLP encode the transaction following eip 2718
+                let mut buf = BytesMut::with_capacity(transaction.encode_2718_len());
+                transaction.encode_2718(&mut buf);
+                let transaction_rlp_bytes = buf.freeze();
+                evm.context.evm.env.tx.scroll.rlp_bytes = Some(transaction_rlp_bytes.into());
+            }
+            evm.context.evm.env.tx.scroll.is_l1_msg = transaction.is_l1_message();
+
+            // execute the transaction and commit the result to the database
+            let ResultAndState { result, state } =
+                evm.transact().map_err(|err| BlockValidationError::EVM {
+                    hash: transaction.recalculate_hash(),
+                    error: Box::new(err.map_db_err(|e| e.into())),
+                })?;
+            evm.db_mut().commit(state);
+
+            let l1_fee = if transaction.is_l1_message() {
+                U256::ZERO
+            } else {
+                // compute l1 fee for all non-l1 transaction
+                let l1_block_info =
+                    evm.context.evm.inner.l1_block_info.as_ref().ok_or_else(|| {
+                        ScrollBlockExecutionError::l1_fee("missing l1 block info")
+                    })?;
+                let transaction_rlp_bytes =
+                    evm.context.evm.env.tx.scroll.rlp_bytes.as_ref().ok_or_else(|| {
+                        ScrollBlockExecutionError::l1_fee("missing transaction rlp bytes")
+                    })?;
```

Comment:
Given that we set this in lines 130 - 134 maybe it's fine to use `expect("this field is set on lines 130 - 134")` instead of `ok_or_else(..)`. What do you think?

Reply:
I did start out with an `expect`, but decided to switch, the reasoning being that we provide a `&mut Evm` to `revm` for the call to transact, so (theoretically) the `revm` implementation could decide to `take` the RLP bytes. If you think it's ok to switch to `expect` I'm happy to update.

Reply:
Yes, I think it's fine to use `expect` here. `revm` is deterministic and we know based on the logic that this will never happen.

Reply:
same for the `l1_block_info` then? this should always be set in the `load_accounts` hook call of the handler.